pkg/storage/stores/shipper/downloads/index_set.go

```diff
@@ -0,0 +1,421 @@
+		return err
+	}
+
+	for _, fileInfo := range filesInfo {
+		if fileInfo.IsDir() {
+			continue
+		}
+
+		fullPath := filepath.Join(t.cacheLocation, fileInfo.Name())
+		// if we fail to open a boltdb file, lets skip it and let sync operation re-download the file from storage.
+		boltdb, err := shipper_util.SafeOpenBoltdbFile(fullPath)
+		if err != nil {
+			level.Error(util_log.Logger).Log("msg", fmt.Sprintf("failed to open existing boltdb file %s, removing the file and continuing without it to let the sync operation catch up", fullPath), "err", err)
+			// Sometimes files get corrupted when the process gets killed in the middle of a download operation which causes boltdb client to panic.
+			// We already recover the panic but the lock on the file is not released by boltdb client which causes the reopening of the file to fail when the sync operation tries it.
+			// We want to remove the file failing to open to get rid of the lock.
+			if err := os.Remove(fullPath); err != nil {
+				level.Error(util_log.Logger).Log("msg", fmt.Sprintf("failed to remove boltdb file %s which failed to open", fullPath))
+			}
+			continue
+		}
+
+		t.dbs[fileInfo.Name()] = boltdb
+	}
+
+	level.Debug(logger).Log("msg", fmt.Sprintf("opened %d local files, now starting sync operation", len(t.dbs)))
+
+	// sync the table to get new files and remove the deleted ones from storage.
+	err = t.sync(ctx, false)
+	if err != nil {
+		return
+	}
+
+	duration := time.Since(startTime).Seconds()
+	t.metrics.tablesDownloadDurationSeconds.add(t.tableName, duration)
+
+	level.Debug(logger).Log("msg", "finished syncing files")
+
+	return
+}
+
+// Close Closes references to all the dbs.
+func (t *indexSet) Close() {
+	// stop the initialization if it is still ongoing.
+	t.cancelFunc()
+
+	err := t.dbsMtx.lock(context.Background())
+	if err != nil {
+		level.Error(t.logger).Log("msg", "failed to acquire lock for closing dbs", "err", err)
+		return
+	}
+	defer t.dbsMtx.unlock()
+
+	for name, db := range t.dbs {
+		if err := db.Close(); err != nil {
+			level.Error(t.logger).Log("msg", fmt.Sprintf("failed to close file %s", name), "err", err)
+		}
+	}
+
+	t.dbs = map[string]*bbolt.DB{}
+}
+
+// MultiQueries runs multiple queries without having to take lock multiple times for each query.
+func (t *indexSet) MultiQueries(ctx context.Context, queries []chunk.IndexQuery, callback chunk_util.Callback) error {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return err
+	}
+
+	userIDBytes := []byte(userID)
+
+	err = t.dbsMtx.rLock(ctx)
+	if err != nil {
+		return err
+	}
+	defer t.dbsMtx.rUnlock()
+
+	if t.err != nil {
+		return t.err
+	}
+
+	t.lastUsedAt = time.Now()
+
+	logger := util_log.WithContext(ctx, t.logger)
+	level.Debug(logger).Log("table-name", t.tableName, "query-count", len(queries))
+
+	for name, db := range t.dbs {
+		err := db.View(func(tx *bbolt.Tx) error {
+			bucket := tx.Bucket(userIDBytes)
+			if bucket == nil {
+				bucket = tx.Bucket(defaultBucketName)
```

Comment:
so this is for format1 and 3. and above for format2 ?

Reply:
yeah, correct.